docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,840 @@
+if my CMP is designed to handle Kustomize projects which contain Helm charts, I might have the CMP accept an
+`ignore-helm-charts` parameter to avoid announciming parameters for those charts.
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+spec:
+  source:
+    plugin:
+      parameters:
+      - name: ignore-helm-charts
+        array: [chart-a, chart-b]
+```
+
+#### How will the CMP know what parameter values are set?
+
+Users persist parameter values in an Application's `spec.source.plugin.parameters` list.
+
+Each parameter has a `name` and a value stored in the `string`, `array`, or `map` field, according to the parameter's 
+collectionType. The name should match the name of some parameter announced by the CMP. (But 
+the user can set any parameter name, so it's the CMP's job to ignore invalid parameters.)
+
+This example is for a hypothetical Helm CMP. This CMP accepts a `values` and a `values-files` parameter.
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+spec:
+  source:
+    repoURL: https://github.com/argoproj/argocd-example-apps.git
+    plugin:
+      parameters:
+        - name: values
+          string: >-
+            resources:
+              cpu: 100m
+              memory: 128Mi
+        - name: values-files
+          array: [values.yaml]
+        - name: helm-parameters
+          map: 
+            image.repository: my.company.com/gcr-proxy/heptio-images/ks-guestbook-demo
+            image.tag: "0.1"
+```
+
+When Argo CD generates manifests (for example, when the user clicks "Hard Refresh" in the UI), Argo CD will send these
+parameters to the CMP as JSON (using the equivalent structure to what's shown above) on an environment variable called
+`ARGOCD_APP_PARAMETERS`.
+
+```shell
+echo "$ARGOCD_APP_PARAMETERS" | jq
+```
+
+That command, when run by a CMP with the above Application manifest, will print the following:
+
+```json
+[
+  {
+    "name": "values",
+    "string": "resources:\n  cpu: 100m\n  memory: 128Mi"
+  },
+  {
+    "name": "values-files",
+    "array": ["values.yaml"]
+  },
+  {
+    "name": "helm-parameters",
+    "map": {
+      "image.repository": "my.company.com/gcr-proxy/heptio-images/ks-guestbook-demo",
+      "image.tag": "0.1"
+    }
+  }
+]
+```
+
+Another way the CMP can access parameters is via environment variables. For example:
+
+```shell
+echo "$VALUES" > /tmp/values.yaml
+helm template --values /tmp/values.yaml .
+```
+
+Environment variable names are set according to these rules:
+
+1. If a parameter is a `string`, the format is `escaped({name})` (`escaped` is defined below).
+2. If a parameter is an `array`, the format is `escaped({name}_{index})` (where the first index is 0).
+3. If a parameter is a `map`, the format is `escaped({name}_{key})`.
+4. If an escaped env var name matches one in the [build environment](https://argo-cd-docs.readthedocs.io/en/latest/user-guide/build-environment/),
```

Comment:
An escaped env var could also match one of the shell's env vars, like `PATH`.  Is that a concern from a security point of view?

Reply:
Excellent catch. Uncommenting the line below does cause the command to fail:

```go
func main() {
	cmd := exec.Command("sh", "-c", "ls")
	//cmd.Env = []string{"PATH="}
	out, err := cmd.Output()
	if err != nil {
		panic(fmt.Sprintf("failed on output: %s", err))
	}
	fmt.Println(string(out))
}
```

What would you think about prefixing all env vars with `PARAM_`?

Reply:
That's probably the easiest way to prevent the problem.

Reply:
Updated. Thanks again for catching that!